frontend/catalyst/jax_primitives.py

```diff
@@ -42,6 +42,11 @@
 from jaxlib.mlir.dialects.scf import ConditionOp, ForOp, IfOp, WhileOp, YieldOp
 from jaxlib.mlir.dialects.stablehlo import ConstantOp as StableHLOConstantOp
 from jaxlib.mlir.dialects.stablehlo import ConvertOp as StableHLOConvertOp
+from mlir_quantum.dialects._transform_ops_gen import (
```

Comment:
They are always generated like that, if desired a pass implementer can create a file `transform.py` which then imports everything from the inc gen file. Worth double checking if it exists.

Reply:
mlir does build the transform dialect's python bindings but not as a `transform.py` file, but as a subdirectory `transform` https://github.com/llvm/llvm-project/blob/59e07f34bdbd463254b0b2744bd0a587d2cd6438/mlir/python/CMakeLists.txt#L179

However its `__init__.py` has an issue when we try to just `from mlir_quantum.dialects.transform import (...)` here:
```
Traceback (most recent call last):
  File "/home/paul.wang/catalyst/frontend/not_a_test_play.py", line 6, in <module>
    import catalyst
  File "/home/paul.wang/catalyst/frontend/catalyst/__init__.py", line 84, in <module>
    from catalyst import debug, logging
  File "/home/paul.wang/catalyst/frontend/catalyst/debug/__init__.py", line 17, in <module>
    from catalyst.debug.callback import callback
  File "/home/paul.wang/catalyst/frontend/catalyst/debug/callback.py", line 16, in <module>
    from catalyst.api_extensions.callbacks import base_callback
  File "/home/paul.wang/catalyst/frontend/catalyst/api_extensions/__init__.py", line 19, in <module>
    from catalyst.api_extensions.callbacks import accelerate, pure_callback
  File "/home/paul.wang/catalyst/frontend/catalyst/api_extensions/callbacks.py", line 40, in <module>
    from catalyst.jax_primitives import python_callback_p
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_primitives.py", line 50, in <module>
    from mlir_quantum.dialects.transform import (
  File "/home/paul.wang/catalyst/frontend/catalyst/../../mlir/build/python_packages/quantum/mlir_quantum/dialects/transform/__init__.py", line 8, in <module>
    from ..._mlir_libs._mlirDialectsTransform import *
ModuleNotFoundError: No module named 'mlir_quantum._mlir_libs._mlirDialectsTransform'
```

The `mlir_quantum._mlir_libs._mlirDialectsTransform` does exist as a `.cpython.so` file; I'm not sure what to do regarding this, since all of these are auto generated from upstream milr when we build. 

I have no objections to using a private here: as long as we build milr before running he frontend (which of course is always the case) this private will always be available.

Reply:
The problem you are encountering here is because we only ship the pure python part of the dialect bindings. Many dialects will also include a C++ extension with additional logic, which we don't want to worry about it because it generally leads to issues importing MLIR shared libraries from us and from jax.

The way we work with this issue is only importing bindings from JAX, from our own dialect bindings where we ignore the import of the C++ extension, or use a couple of additional hard-coded ones in the repo: https://github.com/PennyLaneAI/catalyst/blob/main/frontend/catalyst/utils/extra_bindings.py

I would say you can safely ignore this problem and stick with the private file 👍